### PR TITLE
Enable Ark to store backups/etc. under a prefix within an object storage bucket

### LINKED ISSUE #167: [Feature] Support storing Ark backups within a sub-dir of object storage buckets
Right now Ark only supports storing backups within the root of an object storage bucket. It would be nice to support storing them within sub-dirs as well.

### PR DESCRIPTION
**Pull request by skriss**
Fixes #167 

I'd still like to add some more unit tests within `pkg/persistence`, but other than that this should be ready for a solid review.

**NOTE**: this PR includes a small change to the `ObjectStore` plugin interface (adding a `prefix` parameter to the `ListCommonPrefixes` method).

### CODE REVIEW

**Comment by skriss** [comment]
I started on some unit test updates, but before I go further I would like to get folks' input on the last commit I pushed - which demonstrates a test pattern I'd potentially like to use here, and in some other spots across the code base.

The change is to use a *test fake* instead of a *mock* for `pkg/cloudprovider/ObjectStore`, which is a dependency of the type we're testing (`pkg/persistence/objectBackupStore`).  The test fake is hand-implemented in `pkg/cloudprovider/in_memory_object_store.go`.  The implementation should be semantically ~equivalent to a real cloud-provider implementation of `cloudprovider.ObjectStore`, but it stores all of its data in-proc in a standard golang map.  It also exports all of its internals and provides some helper methods, which make using it within the context of a unit test easier.

I think mocks have their place, and I'm not proposing eliminating them from our toolkit.  But I do think that, in a lot of cases, using a well-crafted fake can reduce the complexity and cognitive overhead associated with tests.  Rather than having to set up mock expectations every time we want to use one, the expectations about how an interface implementation **should** work are embedded in the test fake implementation itself.

Would like everyone's input on this before deciding whether to proceed further in this direction. 
cc @nrb @carlisia @ncdc @wwitzel3

**Comment by wwitzel3** [comment]
I think this is a great approach. I've used this same approach for storage backends. An in-memory fake that satisfies the interface made writing unit tests much less painful.

**Comment by skriss** [comment]
alright, I went ahead and updated the rest of the tests using the fake, and added scenarios for prefixes, so you can now see what it looks like across the board.  Still open to feedback.

**Comment by skriss** [comment]
Will squash before merge!

File: pkg/cloudprovider/aws/object_store.go [hunk 1]
```diff
@@ -116,7 +116,7 @@ func (o *objectStore) Init(config map[string]string) error {
 	return nil
 }
 
-func (o *objectStore) PutObject(bucket string, key string, body io.Reader) error {
```
Comment (skriss) [inline pkg/cloudprovider/aws/object_store.go]: this is an interesting line of code

File: pkg/persistence/object_store_test.go [hunk 2]
```diff
@@ -32,109 +31,190 @@ import (
...
+		prefix:      prefix,
 	}
+}
 
-	return backupStore, objectStore, bucket, backup
+func TestListBackups(t *testing.T) {
+	tests := []struct {
+		name        string
+		prefix      string
+		storageData map[string][]byte
```
[the remark this answers is not in the document]
Comment (skriss) [reply pkg/persistence/object_store_test.go]: `bucketData` makes sense, will update

**Comment by skriss** [comment]
@carlisia @nrb gentle reminder that this PR is ready for a full review.

**Comment by carlisia** [comment]
Ready for squashing!

**Comment by skriss** [comment]
annnnd now all of a sudden a test is failing. looking into...

**Comment by skriss** [comment]
had a flaky test due to comparing things that could be ordered differently. should be fixed now.

**Comment by carlisia** [comment]
Still looks good. Squish away.

**Comment by skriss** [comment]
it's as squashed as I'd like it to be

**Pull request merged**